Google.Api.Generator/Utils/ResourcePattern.cs

```diff
@@ -16,107 +16,154 @@
+                    parameterNames.Add(p);
+                    if (closePos == segment.Length - 1)
                     {
-                        sb.Append(c);
+                        break;
                     }
+                    Check(segment.Length >= closePos + 3, "unexpected end of segment.");
+                    char sep = segment[closePos + 1];
+                    Check(sep == '_' || sep == '-' || sep == '.' || sep == '~', $"invalid separator character '{sep}'.");
+                    separators.Add(sep);
+                    pos = closePos + 2;
                 }
-                if (!inSep)
-                {
-                    throw new ArgumentException($"Segment '{segment}' is ill-formed; missing final '}}'.");
-                }
-                separators.Add(sb.ToString());
-                if (separators.Any(x => x != "") && parameterNames.Zip(parameterNamesWithSuffix, (a, b) => (a, b)).Any(x => x.a != x.b))
+                if (parameterNames.Count > 1)
                 {
-                    throw new ArgumentException($"Segment '{segment}' is ill-formed; multi-id segment may not use '=*' or '=**' suffix.");
+                    Check(parameterNamesWithSuffix.All(p => !p.EndsWith("=**")), "'=**' suffix not valid in a multi-ID segment.");
                 }
                 Separators = separators;
                 ParameterNames = parameterNames;
                 ParameterNamesWithSuffix = parameterNamesWithSuffix;
-            }
 
-            // Seperators include pre- and post- non-parameter-names; there will always be one more seperator than parameter-name.
-            public IReadOnlyList<string> Separators { get; }
-            public IReadOnlyList<string> ParameterNames { get; }
-            public IReadOnlyList<string> ParameterNamesWithSuffix { get; }
-            public int ParameterCount => ParameterNames.Count;
-            public bool IsComplex => ParameterCount > 1 || (ParameterCount == 1 && Separators.Any(x => x != ""));
+                void Check(bool cond, string msg)
+                {
+                    if (!cond)
+                    {
+                        throw new ArgumentException($"Segment '{segment}' is ill-formed; {msg}");
+                    }
+                }
+            }
 
-            public string Expand(IEnumerable<string> parameters) => Separators[0] + string.Join("", parameters.Zip(Separators.Skip(1), (p, s) => p + s));
+            // Separators are between parameter-names; there is always one less separator than parameter-names.
+            public override IReadOnlyList<char> Separators { get; }
+            public override IReadOnlyList<string> ParameterNames { get; }
+            public override IReadOnlyList<string> ParameterNamesWithSuffix { get; }
+            public override string PathTemplateString => $"{{{(ParameterCount == 1 ? ParameterNamesWithSuffix[0] : string.Join('_', ParameterNames))}}}";
```

Comment:
A comment here would be useful. It took me a while to work out what's going on here. (I understand now.)

Reply:
done